File: 2024/python/days/day_4.py
from pathlib import Path

from utils.misc import timer
from utils.grid import make_grid
from utils.read import read_input

example = """MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX"""

def add(p1, p2):
    return p1[0] + p2[0], p1[1] + p2[1]

def mul(p1, n):
    return p1[0] * n, p1[1] * n

add((0,1), (1,1))
mul((1,1), 3)

def part1(input):
    XMAS = ['X', 'M', 'A', 'S']
    grid = make_grid(input)
    dirs = [(0, -1),   # Up
            (0, 1),    # Down
            (-1, 0),   # Left
            (1, 0),    # Right
            (-1, -1),  # Up-Left
            (1, -1),   # Up-Right
            (-1, 1),  # Down-Left
            (1, 1)]    # Down-Right
    point = (0, 0)
    answer = 0
    while point[1] <= grid.y_max:
        for d in dirs:
            path = [point, add(point, d), add(point, mul(d, 2)), add(point, mul(d, 3))]
            letters = list(map(lambda p: grid.at(p[0], p[1]) if grid.in_bounds(p[0], p[1]) else None, path))
            if letters == XMAS:
                answer += 1
        point = (point[0] + 1 if point[0] < grid.x_max else 0, point[1] if point[0] < grid.x_max else point[1] + 1)
    print(f'Pt1::ans: {answer}')

example_2 = """
.M.S......
..A..MSMS.
.M.S.MAA..
..A.ASMSM.
.M.S.M....
..........
S.S.S.S.S.
.A.A.A.A..
M.M.M.M.M.
..........""".strip()

# mas_0 = """
# M.S
# .A.
# S.M
# """.strip()

mas_1 = """
M.S
.A.
M.S
""".strip()

mas_2 = """
M.M
.A.
S.S
""".strip()

mas_3 = """
S.M
.A.
S.M
""".strip()

# mas_4 = """
# S.M
# .A.
# M.S
# """.strip()

mas_4 = """
S.S
.A.
M.M
""".strip()

def part2(input):
    grid = make_grid(input)
    x_mases = [list(''.join(mas.split('\n'))) for mas in [mas_1, mas_2, mas_3, mas_4]]
    point = (0, 0)
    answer = 0
    while point[1] <= grid.y_max:
        sub_grid = [(0, 0), (1, 0), (2, 0),
                    (0, 1), (1, 1), (2, 1),
                    (0, 2), (1, 2), (2, 2)]
        sub_grid_points = [add(point, p) for p in sub_grid]
        sub_grid_at_point = [grid.at(p[0], p[1]) if grid.in_bounds(p[0], p[1]) else None for p in sub_grid_points]

        for mas in x_mases:
            if sub_grid_at_point[::2] == mas[::2]:
                answer += 1
                break
        point = (point[0] + 1 if point[0] < grid.x_max else 0, point[1] if point[0] < grid.x_max else point[1] + 1)
    print(f'Pt2::ans: {answer}')

def run():
    day = Path(__file__).name.split('.')[0].split('_')[-1]
    input = read_input(day)
    with timer():
        part1(example)

    with timer():
        part1(input)

    with timer():
        part2(example_2)

    with timer():
        part2(input)

if __name__ == "__main__":
    run()

File: 2024/cpp23/days/day_4.cpp
#include <string>
#include <utils/misc.hpp>

std::string example_2(R"(.M.S......
..A..MSMS.
.M.S.MAA..
..A.ASMSM.
.M.S.M....
..........
S.S.S.S.S.
.A.A.A.A..
M.M.M.M.M.
..........)");

void part_i(const std::string &input) {
    std::println("Part i: {}", "Not implemented");
}


void part_ii(const std::string &input) {
//    std::string_view grid(input);
    std::array<char, 5> mas_1{'M', 'S', 'A', 'M', 'S'};
    std::array<char, 5> mas_2{'M', 'M', 'A', 'S', 'S'};
    std::array<char, 5> mas_3{'S', 'M', 'A', 'S', 'M'};
    std::array<char, 5> mas_4{'S', 'S', 'A', 'M', 'M'};
    std::array<std::array<char, 5>, 4> mases_to_check{mas_1, mas_2, mas_3, mas_4};

    size_t block_len = input.find('\n') + 1;
    std::array<size_t, 5> idxs_to_check{0, 2, block_len + 1, 2 * block_len, 2 * block_len + 2};
    int answer = 0;
    while (idxs_to_check[4] < input.length()) {
        for (const auto &mas: mases_to_check) {
            if (mas[0] == input[idxs_to_check[0]] &&
                mas[1] == input[idxs_to_check[1]] &&
                mas[2] == input[idxs_to_check[2]] &&
                mas[3] == input[idxs_to_check[3]] &&
                mas[4] == input[idxs_to_check[4]]
            ) {
                answer += 1;
            }
        }
        bool go_to_next_line = (idxs_to_check[1] % block_len) == (block_len - 2);
        for (auto &n: idxs_to_check) {
            if (go_to_next_line) {
                n += 2;
            } else {
                n +=1;
            }
        }
    }

    std::println("Part ii: {}", answer);
}

int main() {
    auto input = Utils::read_input(4);
//    Utils::with_timer("Parse input",  [&input](){ parse_input(input); });

//    Utils::with_timer("Part i (example)", []() {
//        part_i(example);
//    });

//    Utils::with_timer("Part i (input)", [&input]() {
//        part_i(input);
//    });

    Utils::with_timer("Part ii (example)", []() {
        part_ii(example_2);
    });

    Utils::with_timer("Part ii (input)", [&input]() {
        part_ii(input);
    });
}